sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -230,9 +231,19 @@ case class Union(children: Seq[LogicalPlan]) extends LogicalPlan {
   }
 
   // updating nullability to make all the children consistent
-  override def output: Seq[Attribute] =
-    children.map(_.output).transpose.map(attrs =>
-      attrs.head.withNullability(attrs.exists(_.nullable)))
+  override def output: Seq[Attribute] = {
+    children.map(_.output).transpose.map { attrs =>
+      val firstAttr = attrs.head
+      val outAttr = try {
+        firstAttr.withDataType(attrs.map(_.dataType).reduce(StructType.merge))
+      } catch {
+        // This should never happen, since `output` is called only when the plan is `resolved`, so
```

Comment:
I understand that this is a fallback logic. But, if this should never happen, we should use more explicit ways. We need to find bugs eagerly. For example, `assert()`, `logWarning`, or `not using try` at all.

Reply:
thanks for the comment @dongjoon-hyun. I propose to add a `logError` in order to ask users to report this error condition and go on with this fallback logic, in order to ensure that the behavior is the same as before this patch in unexpected conditions (so no behavior change) but at the same time we can detect this problem. Do you agree?